### PR TITLE
[Perf] Skip logits and sampling for unfinished prefills

### PR DESCRIPTION
**Pull request by jesse996**
## Purpose

Model Runner V2 currently produces one sampling-logit row for every request in the common non-speculative path, including chunked-prefill requests whose prefill does not finish in the current step.

Although these sampled tokens are discarded later by setting `num_sampled` to zero, the model has already performed the vocabulary projection and the sampler has already applied logits processors and sampling kernels. This introduces unnecessary GPU work, especially for workloads with multiple concurrent long prompts and a small `max_num_batched_tokens`.

This PR avoids that work by:

- Determining whether each request finishes prefill before constructing the logits batch.
- Assigning zero logits rows to unfinished chunked-prefill requests.
- Compacting mixed prefill/decode batches so output projection and sampling run only for requests that can produce an output token.
- Short-circuiting `compute_logits` and sampling when the entire batch consists of unfinished prefill chunks.
- Expanding compact sampled-token and NaN-count results back to the original per-request layout with a small Triton kernel, preserving existing downstream interfaces.
- Preserving the existing pure-decode fast path.

The change is limited to Model Runner V2's common non-speculative path. Speculative decoding remains unchanged. LoRA batches retain the existing dense logits layout because the per-request mapping is reused by prompt-logprob
projection.

This should reduce unnecessary vocabulary-projection and sampling work for chunked-prefill-heavy serving workloads without changing generated outputs.

### start server
```
VLLM_USE_V2_MODEL_RUNNER=1 \
vllm serve Qwen3.5-4B \
    --served-model-name Qwen3.5-4B \
    --host 0.0.0.0 \
    --port 8000 \
    --enable-chunked-prefill \
    --max-num-batched-tokens 2048 \
    --max-num-seqs 32 \
    --gpu-memory-utilization 0.90 \
    --disable-log-stats
```
### vllm bench
```
vllm bench serve \
    --backend vllm \
    --host 127.0.0.1 \
    --port 8000 \
    --endpoint /v1/completions \
    --model Qwen3.5-4B \
    --tokenizer Qwen3.5-4B \
    --dataset-name random \
    --random-input-len 8192 \
    --random-output-len 32 \
    --random-range-ratio 0 \
    --ignore-eos \
    --request-rate inf \
    --max-concurrency 32 \
    --num-warmups 16 \
    --num-prompts 256 \
    --seed 0 \
    --percentile-metrics ttft,tpot,itl,e2el \
    --metric-percentiles 50,90,99
```
### before
IMG_3103

### after
IMG_3104

- Output token throughput increased from **174.72 tok/s** to **191.97 tok/s**, an improvement of **9.87%**.
- Mean TTFT decreased from **4311.56 ms** to **3904.83 ms**, a reduction of **9.43%**.
- Mean TPOT decreased from **43.97 ms** to **40.15 ms**, a reduction of **8.69%**.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by jesse996** [comment]
@yewentao256 pls help to review this，thank you very much！

**Review by yewentao256 — COMMENTED** [review]
Thanks for the work! Please add a lm_eval test to make sure we don't hurt acc here

**Comment by jesse996** [comment]
> Thanks for the work! Please add a lm_eval test to make sure we don't hurt acc here

Thanks for the reply! lm_eval is not available in my current environment, so I used evalscope to run an accuracy regression test instead.
```
evalscope eval --model /nas/disk1/Qwen3.5-4B --dataset-hub Local --generation-config {"max_tokens":10240,"top_p":0.8,"top_k":20,"temperature": 0.7,"n":1,"timeout": 60,"stream":true,"extra_body":{"chat_template_kwargs":{"enable_thinking":false}}} --eval-type openai_api --eval-batch-size 128 --api-key EMPTY --api-url http://127.0.0.1:30019/v1 --ignore-errors --datasets ceval --dataset-args {"ceval":{"local_path":"/model/modelscope-cache/datasets/ceval"}}
```

I didn’t observe accuracy regression. In fact, the accuracy improved slightly,maybe within the margin of error

## before
IMG_3138

## after
IMG_3139

File: vllm/v1/worker/gpu/sample/sampler.py [hunk 1]
```diff
@@ -129,6 +160,18 @@ def __call__(
...
+            input_batch.num_draft_tokens_per_req is None
+            and sampled.shape[0] != input_batch.num_reqs
+        ):
+            sampled = _expand_compact_values(
+                sampled, input_batch.cu_num_logits, fill_value=-1
+            )
+            if num_nans is not None:
+                num_nans = _expand_compact_values(
+                    num_nans, input_batch.cu_num_logits, fill_value=0
+                )
```
Comment (yewentao256) [inline vllm/v1/worker/gpu/sample/sampler.py:210]: Will this be warmed up like https://github.com/vllm-project/vllm/pull/49903?

Comment (jesse996) [reply vllm/v1/worker/gpu/sample/sampler.py:210]: I’ve added an warmup in the V2 warmup_kernels() path, similar to #49903. It precompiles all runtime variants of the new expansion kernel.

**Review by yewentao256 — COMMENTED** [review]
Generally LGTM, also CC @njhill

**Comment by malaiwah** [comment]
I audited this optimization against an MRV2 native-MTP path
(`local-inference-lab/vllm@e2666d9a65`, GLM-5.2, MTP3).

One useful clarification: this PR's no-scheduled-drafts branch should already
skip target LM-head/sampling work for a pure unfinished-prefill batch even when
an MTP speculator is configured. The remaining speculative gaps are narrower:

1. A mixed batch with decode rows and unfinished-prefill rows enters the global
   draft-token branch, which still gives every request a target bonus-logit row.
2. The MTP drafter must run its prefill forward to hydrate draft KV, but it then
   computes/samples the first proposal and runs the remaining autoregressive
   MTP steps for unfinished rows. The scheduler explicitly discards those
   proposals.

I opened a source-backed follow-up with the exact call sites, a hydrate-only
design, risks and GPU A/B contract:
https://github.com/local-inference-lab/vllm/issues/272

For the 753B GLM shape I would expect low-single-digit benefit under concurrent
long prefill, not assume this PR's Qwen3.5-4B headline. No GLM performance claim
is being made before the matched GPU test.

**Review by yewentao256 — COMMENTED** [review]
Genearally LGTM, thanks for the work!
Also CC @njhill

**Comment by njhill** [comment]
Thanks @jesse996 I'll try to look at this soon. I had actually experimented with a similar change previously, but it adds nontrivial complexity to the model runner and sampler, so I'd like to look at it a bit more carefully.

also cc @WoosukKwon

**Comment by WoosukKwon** [comment]
Thanks for the PR. I think this optimization is valuable, but the current code change is a bit too complex. Can we further simplify this?

**Comment by jesse996** [comment]
> Thanks for the PR. I think this optimization is valuable, but the current code change is a bit too complex. Can we further simplify this?

Thanks for the reply! l have simplify the code and rerun my test,pls take a look~

**Comment by jesse996** [comment]
@WoosukKwon could you please take a look？
